File: lib/pdf-generator.ts
// frontend/lib/generateCoursesPDF.ts
"use client";

import { jsPDF } from "jspdf";
import autoTable from "jspdf-autotable";
import type { Course } from "@/types";

// Function to generate a PDF from selected courses
export const generateCoursesPDF = (courses: Course[], userName = "Student") => {
  // Create a new PDF document
  const doc = new jsPDF();

  // Add title
  doc.setFontSize(20);
  doc.setTextColor(39, 174, 96); // Green color
  doc.text("EduPathway - Selected Courses", 14, 22);

  // Add subtitle with date
  doc.setFontSize(12);
  doc.setTextColor(100, 100, 100);
  const today = new Date().toLocaleDateString();
  doc.text(`Generated for: ${userName} | Date: ${today}`, 14, 30);

  // Add description
  doc.setFontSize(10);
  doc.setTextColor(80, 80, 80);
  doc.text(
    "This document contains your selected university courses based on your academic profile and interests.",
    14,
    38,
  );

  // Create table header and data
  const tableColumn = ["Course Code", "Title", "University", "Points", "Description"];
  const tableRows = courses.map((course) => [
    course.code,
    course.name,
    course.university_name,
    course.minimum_grade,
    course.description.substring(0, 60) + (course.description.length > 60 ? "..." : ""),
  ]);

  // Generate the table
  autoTable(doc, {
    head: [tableColumn],
    body: tableRows,
    startY: 45,
    styles: { fontSize: 9, cellPadding: 3 },
    headStyles: { fillColor: [39, 174, 96], textColor: [255, 255, 255] },
    alternateRowStyles: { fillColor: [240, 240, 240] },
    margin: { top: 45 },
  });

  // Add course details section
  let yPos = (doc as any).lastAutoTable.finalY + 15;

  doc.setFontSize(14);
  doc.setTextColor(39, 174, 96);
  doc.text("Course Details", 14, yPos);

  yPos += 8;

  // Add detailed information for each course
  courses.forEach((course, index) => {
    // Check if we need a new page
    if (yPos > 250) {
      doc.addPage();
      yPos = 20;
    }

    doc.setFontSize(12);
    doc.setTextColor(50, 50, 50);
    doc.text(`${index + 1}. ${course.name} (${course.code})`, 14, yPos);

    yPos += 6;

    doc.setFontSize(10);
    doc.setTextColor(80, 80, 80);
    doc.text(`University: ${course.university_name}`, 18, yPos);

    yPos += 5;

    doc.text(`Required Points: ${course.minimum_grade}`, 18, yPos);

    yPos += 5;

    // Split long descriptions into multiple lines
    const descLines = doc.splitTextToSize(course.description, 170);
    doc.text(descLines, 18, yPos);

    yPos += descLines.length * 5 + 8;

    // Add additional course details
    if (course.duration_years) {
      doc.text(`Duration: ${course.duration_years} years`, 18, yPos);
      yPos += 5;
    }

    if (course.applicationDeadline) {
      doc.text(`Application Deadline: ${course.applicationDeadline}`, 18, yPos);
      yPos += 5;
    }

    if (course.required_subjects && course.required_subjects.length > 0) {
      const subjects = course.required_subjects.map((s) => s.subject.name).join(", ");
      doc.text(`Required Subjects: ${subjects}`, 18, yPos);
      yPos += 5;
    }

    if (course.career_prospects) {
      doc.text(`Career Paths: ${course.career_prospects}`, 18, yPos);
      yPos += 5;
    }

    // Add some space between courses
    yPos += 8;
  });

  // Add footer
  const pageCount = doc.getNumberOfPages();
  for (let i = 1; i <= pageCount; i++) {
    doc.setPage(i);
    doc.setFontSize(8);
    doc.setTextColor(150, 150, 150);
    doc.text(
      "EduPathway - Find your perfect university course | www.edupathway.co.ke",
      14,
      doc.internal.pageSize.height - 10,
    );
    doc.text(`Page ${i} of ${pageCount}`, doc.internal.pageSize.width - 25, doc.internal.pageSize.height - 10);
  }

  // Save the PDF
  doc.save("EduPathway-Selected-Courses.pdf");
};

File: backend/EDUHUB/utils/pdf_generator.py
# backend/utils/pdf_generator.py
from io import BytesIO
from reportlab.pdfgen import canvas
from reportlab.lib.pagesizes import letter
from reportlab.lib import colors
from reportlab.platypus import Table, TableStyle

def generate_pdf(courses, user_name="Student"):
    buffer = BytesIO()
    p = canvas.Canvas(buffer, pagesize=letter)
    width, height = letter
    
    # Add title
    p.setFont("Helvetica-Bold", 20)
    p.setFillColor(colors.green)
    p.drawString(72, height - 72, "EduPathway - Selected Courses")
    
    # Add subtitle with date
    p.setFont("Helvetica", 12)
    p.setFillColor(colors.darkgrey)
    from datetime import date
    today = date.today().strftime("%B %d, %Y")
    p.drawString(72, height - 100, f"Generated for: {user_name} | Date: {today}")
    
    # Add description
    p.setFont("Helvetica", 10)
    p.drawString(72, height - 120, 
                "This document contains your selected university courses based on your academic profile and interests.")
    
    # Create table data
    data = [["Course Code", "Title", "University", "Points"]]
    for course in courses:
        data.append([course.code, course.title, course.university, str(course.points)])
    
    # Create table
    table = Table(data, colWidths=[80, 200, 150, 50])
    table.setStyle(TableStyle([
        ('BACKGROUND', (0, 0), (-1, 0), colors.green),
        ('TEXTCOLOR', (0, 0), (-1, 0), colors.white),
        ('ALIGN', (0, 0), (-1, -1), 'CENTER'),
        ('FONTNAME', (0, 0), (-1, 0), 'Helvetica-Bold'),
        ('BOTTOMPADDING', (0, 0), (-1, 0), 12),
        ('BACKGROUND', (0, 1), (-1, -1), colors.white),
        ('GRID', (0, 0), (-1, -1), 1, colors.black),
    ]))
    
    # Draw table
    table.wrapOn(p, width - 144, height)
    table.drawOn(p, 72, height - 200)
    
    # Add course details
    y_position = height - 250
    p.setFont("Helvetica-Bold", 14)
    p.setFillColor(colors.green)
    p.drawString(72, y_position, "Course Details")
    
    y_position -= 30
    
    for i, course in enumerate(courses):
        if y_position < 100:  # Check if we need a new page
            p.showPage()
            y_position = height - 72
        
        p.setFont("Helvetica-Bold", 12)
        p.setFillColor(colors.black)
        p.drawString(72, y_position, f"{i+1}. {course.title} ({course.code})")
        
        y_position -= 20
        p.setFont("Helvetica", 10)
        p.drawString(90, y_position, f"University: {course.university}")
        
        y_position -= 15
        p.drawString(90, y_position, f"Required Points: {course.points}")
        
        y_position -= 15
        p.drawString(90, y_position, f"Duration: {course.duration}")
        
        y_position -= 15
        p.drawString(90, y_position, f"Start Date: {course.start_date}")
        
        y_position -= 15
        p.drawString(90, y_position, f"Application Deadline: {course.application_deadline}")
        
        y_position -= 30
    
    # Add footer
    p.setFont("Helvetica", 8)
    p.setFillColor(colors.grey)
    p.drawString(72, 40, "EduPathway - Find your perfect university course | www.edupathway.com")
    p.drawString(width - 150, 40, f"Page 1 of 1")
    
    p.showPage()
    p.save()
    
    pdf = buffer.getvalue()
    buffer.close()
    return pdf